src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -338,67 +312,102 @@ public void onClose() {
         logger.debug("closed");
       }
       if (!isClosed) {
-        executorService.execute(new Runnable() {
-          @Override
-          public void run() {
-            onClosed();
-          }
-        });
+        // If the connection tear down was initiated by the higher-layer, isClosed will already
+        // be true. Nothing more to do in that case.
+        executorService.execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                onClosed();
+              }
+            });
       }
     }
 
     @Override
-    public void onError(final WebSocketException e) {
+    public void onError(final Throwable e) {
+      if (e.getCause() != null && e.getCause() instanceof EOFException) {
+        logger.error("WebSocket reached EOF", e);
+      } else {
+        logger.error("WebSocket error", e);
+      }
       executorService.execute(
           new Runnable() {
             @Override
             public void run() {
-              if (e.getCause() != null && e.getCause() instanceof EOFException) {
-                logger.debug("WebSocket reached EOF.");
-              } else {
-                logger.debug("WebSocket error.", e);
-              }
               onClosed();
             }
           });
     }
+  }
 
-    @Override
-    public void onLogMessage(String msg) {
-      if (logger.logsDebug()) {
-        logger.debug("Tubesock: " + msg);
-      }
-    }
+  private static class StringList {
 
-    @Override
-    public void send(String msg) {
-      ws.send(msg);
+    private final AtomicInteger remaining = new AtomicInteger(0);
```

Comment:
Why AtomicInteger?  This class should only be accessed by a single thread, right?  Using AtomicInteger suggests otherwise which makes the code less clear.

Reply:
Done